src/web/arr/trove/chart.arr

```diff
@@ -0,0 +1,1028 @@
+  | bar-chart-series(obj :: BarChartSeries) with:
+    is-single: true,
+    constr: {(): bar-chart-series},
+  | histogram-series(obj :: HistogramSeries) with:
+    is-single: true,
+    constr: {(): histogram-series},
+    method bin-width(self, bin-width :: Number):
+      histogram-series(self.obj.{bin-width: bin-width})
+    end,
+    method max-num-bins(self, max-num-bins :: Number):
+      histogram-series(self.obj.{max-num-bins: max-num-bins})
+    end,
+    method min-num-bins(self, min-num-bins :: Number):
+      histogram-series(self.obj.{min-num-bins: min-num-bins})
+    end,
+    method num-bins(self, num-bins :: Number):
+      histogram-series(self.obj.{min-num-bins: num-bins, max-num-bins: num-bins})
+    end,
+    method value-name(self, value-name :: String):
+      histogram-series(self.obj.{value-name: value-name})
+    end,
+end
+
+fun check-chart-window(p :: ChartWindowObject) -> Nothing block:
+  when (p.width <= 0) or (p.height <= 0):
+    raise('render: width and height must be positive')
+  end
+end
+
+################################################################################
+# METHODS
+################################################################################
+
+title-method = method(self, title :: String):
+  self.constr()(self.obj.{title: title})
+end
+
+width-method = method(self, width :: Number):
+  self.constr()(self.obj.{width: width})
+end
+
+height-method = method(self, height :: Number):
+  self.constr()(self.obj.{height: height})
+end
+
+x-axis-method = method(self, x-axis :: String):
+  self.constr()(self.obj.{x-axis: x-axis})
+end
+
+y-axis-method = method(self, y-axis :: String):
+  self.constr()(self.obj.{y-axis: y-axis})
+end
+
+x-range-method = method(self, x-min :: Number, x-max :: Number) block:
+  when (x-min >= x-max) and (not(unsafe-equal(x-min, 0)) or
+      not(unsafe-equal(x-max, 0))):
+    raise([sprintf: 'x-range: x-min must be strictly less than x-max, ',
+        'except x-min, x-max both equal 0 which will infer ',
+        'appropriate values'])
+  end
+  self.constr()(self.obj.{x-min: x-min, x-max: x-max})
+end
+
+y-range-method = method(self, y-min :: Number, y-max :: Number) block:
+  when (y-min >= y-max) and (not(unsafe-equal(y-min, 0)) or
+      not(unsafe-equal(y-max, 0))):
+    raise([sprintf: 'y-range: y-min must be strictly less than y-max, ',
+        'except y-min, y-max both equal 0 which will infer ',
+        'appropriate values'])
+  end
+  self.constr()(self.obj.{y-min: y-min, y-max: y-max})
+end
+
+
+data ChartWindow:
+  | pie-chart-window(obj :: ChartWindowObject) with:
+    constr: {(): pie-chart-window},
+    title: title-method,
+    width: width-method,
+    height: height-method,
+  | bar-chart-window(obj :: ChartWindowObject) with:
+    constr: {(): bar-chart-window},
+    title: title-method,
+    width: width-method,
+    height: height-method,
+    x-axis: x-axis-method,
+    y-axis: y-axis-method,
+    y-range: y-range-method,
```

Comment:
I'm not exactly sure what is your concern, so let me put all of my thought when I designed this, and hopefully it will address your concern.

1. This is bar-chart, so it can't have `x-range` (if that's your concern)
2. Having both `y-min` and `y-max` 0 will automatically compute the bound for y-axis. Initially, I have `y-min` and `y-max` separately, but then there is no good value for "please compute the bound for me automatically" because 0 could actually mean value 0. But if we combine them to `y-range`, we know that the range [0, 0] doesn't make sense, so we can use that as a default value for "please infer it for me".
3. I don't want to have `range(x-min, x-max, y-min, y-max)` because it's not easy to remember -- is it `range(x-min, x-max, y-min, y-max)` or `range(y-min, y-max, x-min, x-max)` or `range(x-min, y-min, x-max, y-max)`?
4. I want to be able to have a way to "reset" back. For example, I could write:

```
f(plot(...)
  .y-range(..., ...))
```

Whoever (not necessary the one who writes the above code) writing function `f` might want to set `y-range` to the default value and then display it. This interface provides a way to do this nicely.

Reply:
1. No, not my worry
2. That's what I'm worried about -- I'd like a way to say "I only want to specify the max (or min, respectively); please automatically compute the other one"
3. No, I'm not saying that either.  I'm saying having `y-min`, `y-max`, `x-min` and `x-max` as four separate methods gives finer grained control.
4. That's fine, but that's an orthogonal concern to what I'm suggesting.

In some sense, specifying `y-min = +Infinity` is a valid reset for the minimum value, and `y-max = -Infinity` is a valid reset for the maximum.  You could store those values internally as the defaults, then use the four individual methods I suggest above to override them, and then your `reset` method could set things back to the infinities that are the defaults.

Reply:
Note that infinity isn't a value in Pyret.

I agree with Ben's suggesting of multiple methods. Why not start auto-computed for both, and then have the `y-max` and `y-min` methods cause them to be individually set?